x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/SqlParser.java

```diff
@@ -240,12 +273,35 @@ public void enterEveryRule(ParserRuleContext ctx) {
 
         @Override
         public void exitEveryRule(ParserRuleContext ctx) {
-            // Avoid having negative numbers
-            if (depthCounts.containsKey(ctx.getClass().getSimpleName())) {
-                depthCounts.putOrAdd(ctx.getClass().getSimpleName(), (short) 0, (short) -1);
+            if (inDetected(ctx)) {
+                insideIn = false;
             }
+
+            decrementCounter(ctx);
             super.exitEveryRule(ctx);
         }
+
+        ObjectShortHashMap<String> depthCounts() {
+            return depthCounts;
+        }
+
+        private void decrementCounter(ParserRuleContext ctx) {
+            String className = ctx.getClass().getSimpleName();
+            String classNameToDecrement = ENTER_EXIT_RULE_MAPPING.getOrDefault(className, className);
+
+            // Avoid having negative numbers
+            if (depthCounts.containsKey(classNameToDecrement)) {
+                depthCounts.putOrAdd(classNameToDecrement, (short) 0, (short) -1);
```

Comment:
Is the scenario of negative counter even possible?

Reply:
Yes, because currently some rules are not encountered on enter but only on exit.
I'll open an issue soon to track this and investigate more.